### ISSUE TITLE
[Compiler Bug]: useMemo does not provide stable results across renders for specific functions when compiler enabled

### ISSUE THREAD

**Issue by bjnsn** [facebook/react#31462]
### What kind of issue is this?

- [X] React Compiler core (the JS output is incorrect, or your app works incorrectly after optimization)
- [ ] babel-plugin-react-compiler (build issue installing or using the Babel plugin)
- [ ] eslint-plugin-react-compiler (build issue installing or using the eslint plugin)
- [ ] react-compiler-healthcheck (build issue installing or using the healthcheck script)

### Link to repro

https://playground.react.dev/#N4Igzg9grgTgxgUxALhAejQAgMIBsCWcA1vgHYDmmALgBYKYK4IC2CpVmA7vrrpmXBgs2HWvTgFimONHYAdUhkwBDUgBNMAMwjxxEZgAcIpEZiEBaIeoQwFSgMoD6AN2W4oCAIwr1mV+4QAJhUhTCgwMkoAA3CEAFkWCCiufFpMOyxVBkMqAE9MNQQDNkLSOHzlGBhlXIAaajp8sBpoXA1SCA44GlVyBAA6DMwAdVSaHH0DHhsGUmUAIyY1ev8PYO7ehDBB0gUEAA8jGA5CzWUoXA5NKDKqfGNMONyAQQMDAAoASkxgBUxMJRgKgLJgqQQQMBgMwlGxgP7SYxAvxuDzeAC8YTA8US7y+mDRAD5HspaP1qup9F96gBtAC6nwA3Ap4UoOhwgSD6MpwZDodYYFDOFgZIZpjB0qR-kp5lAOPhBWNoF1JmKmZKAcKspwdEQQrINOFIhKpcLERxqasgrUAPq0-GYhD2YFUBC476E4mk8lqSnfADUmE8n0+8JkpCRluCGNiCWYEDd+KJcRJNDJqh9zDxAc8NPpapZwskJAoKV40MsMNs6rDSOpAiErHY9SxVAAkmUGyI7dGsU6Sa6AAyM5nqoRUWCSgA8anwzkwxjwhCIaOAeI9LfbgmE7He+ET-Ew2c+AF8CRIl8hgPXt1RjysUV5L5bPHfkQFAk+H4Fj5O0DPnASarHiAx5AA

### Repro steps

1. Write a `useMemo` function that uses Math.random() AND some other value or function, for example:
-  `const val = useMemo(() => Math.random() + 0, []);`
- `const val = useMemo(() => Math.floor(Math.random()), []);`
2. Cause the component to re-render
3. The resulting value will be re-computed on re-render

In associated repro link, clicking the component increments a counter which triggers re-render. Without compiler it both memoized values are stable; with compiler, only the first is.

### How often does this bug happen?

Every time

### What version of React are you using?

18.3.1

### What version of React Compiler are you using?

19.0.0-beta-8a03594-20241020

**Comment by HarshTripathi1103** [comment]
i guess its the optimization conflict between the react compiler version and useMemo().

**Comment by josephsavona** [comment]
Thanks for posting. In general using functions such as Date.now() or Math.random() in render are not recommended precisely because they can change arbitrarily. Think about `useMemo()` as a performance optimization and not something to take impure code and make it pure. You can read more about this at https://react.dev/reference/rules#components-and-hooks-must-be-pure

That said, we want to make the compiler work with as much existing code as possible. What's happening here is that in the first `useMemo()`, the body is a function call which we memoize. For the second `useMemo()`, the body is an addition expression which we know produces a number. Numbers don't need to be memoized since they can be compared with `===`, so we prune the memoization. Per the Rules of React this is a safe optimization, but obviously in can break code if you're violating the rules.

One thing we should consider here is to teach the compiler about Date.now and Math.random and bail out of compilation if they're accessed in render — that would mean we'd just safely skip code like this and keep it working as-is.

**Comment by bjnsn** [comment]
Hmm. I tend to lean on the eslint to make sure I’m obeying the rules…and it won’t be easy to catch this one…there’s plenty of libraries that are also impure (e.g. seeded random, uid generators) in addition to JS standard methods.

I’ve often seen useMemo used to compute something once on first render…one path would be to create a hook explicitly for that (though obviously people can roll their own)

**Comment by 1057105012** [comment]
> Thanks for posting. In general using functions such as Date.now() or Math.random() in render are not recommended precisely because they can change arbitrarily. Think about `useMemo()` as a performance optimization and not something to take impure code and make it pure. You can read more about this at https://react.dev/reference/rules#components-and-hooks-must-be-pure
> 
> That said, we want to make the compiler work with as much existing code as possible. What's happening here is that in the first `useMemo()`, the body is a function call which we memoize. For the second `useMemo()`, the body is an addition expression which we know produces a number. Numbers don't need to be memoized since they can be compared with `===`, so we prune the memoization. Per the Rules of React this is a safe optimization, but obviously in can break code if you're violating the rules.
> 
> One thing we should consider here is to teach the compiler about Date.now and Math.random and bail out of compilation if they're accessed in render — that would mean we'd just safely skip code like this and keep it working as-is.

I think that manual operations of useMemo and useCallback should be regarded as conscious and skipped during optimization. It does not care about right or wrong at all.

Because the purpose of the compiler is not to implicitly correct all errors, but to optimize the execution efficiency of the code. These are two different concepts.

At the same time, I think the react compiler can consider running statistics and dynamic optimization in the background during debugging, and then generate a react-compiler-config file. According to my superficial experience, it is easier to determine where there are performance bottlenecks based on dynamic information.

**Comment by josephsavona** [comment]
> I think that manual operations of useMemo and useCallback should be regarded as conscious and skipped during optimization. It does not care about right or wrong at all.

Yup, this is basically the compiler does, there is just one specific exception. As a general rule, the compiler will only compile code if it can maintain all the original memoization, so you're guaranteed to have at least as much memoization as what you wrote. The only exception is if we can infer that a memo block produces a primitive value, we currently prune the memoization. Based on this and some similar issues we are likely to remove this exception so that we always retain manual memoization.

**Comment by pratapmajge** [comment]
I’d like to work on this.

**Comment by josephsavona** [comment]
Note that we have implemented best-effort validation against known impure functions like Math.random() and Date.now(). This will be enabled by default in an upcoming release of the linter. You can see a preview of this for the [code from the issue](https://playground.react.dev/#N4Igzg9grgTgxgUxALhAejQAgAIDcCGANgJYAm+ALggHIQCSAtgA6wIBiUAdnBcRJ2DqcASgk6kEMADqcMmAMIk4Aa2KcA5pgoALBJgSEEDMRUwB3YoUKY1cGEZNbdmOEuUvonCjLn5xmADMIeD04CGZ+R3sAWntxSR8sAGVbPQJCKAQARkw-Ukx0zIAmXPtMKDA1TQADCoQAWSMIavNiHUxE3M59ZgoAT0wJJjEJbgH8GBh8PoAaJwQBsG1oQnzOCFM4bT91BAA6ToB1Nu0FcKZLSX1OfAAjQ1I5woQSrZ2EMAPOGQQADyZgqYJAF8FBCKYAlweHxuvU+gBBJhMAAUAEpMMAZJhMHIwBQ7oZcnYIGAwJg4hIYGAsR4BKZnjkALzlMANJrItGYRkAPkw9Uo2j2U3E4TRcwA2gBdVEAbhkNLk61MeIJenwxNJ5JGkjJZiwYWYlxgHW6OKwtygpmIupO0E25yNctNcjgfnMwXcE08+QqVRN2Jd-DxmHFzyKMwA+pKuSyEEl8VQOeieXyBUK8qL0QBqTBZVGomlhOkFIjFGN1RoMCBJrm8-k6dMihicnNZCXSp0K-VuP0WKxa2La6SmovB8W2ezGLxzVkUIR2BxeaPMurxygIZEABll8tN9gosG6AB5SMRcJh+IpiCpGcBOSnZ-PJyZkcRazZMK3UQBfbmua-KMgwATouFDfk8pbZEBDLgSWGQvNBkFFN+R5oKeuDck634gN+QA)
